### PR TITLE
feat: implement MonoVertexRollout controller w/ CRUD and autohealing

### LINKED ISSUE #197: MonoVertexRollout basic CRUD

### PR DESCRIPTION
**Pull request by dpadhiar**
Fixes #197 

### Modifications

Implements the MonoVertexRollout controller based on the ISBServiceRollout controller logic.

MonoVertexRollouts now can create, delete and update child MonoVertex resources. If a child MonoVertex is updated anything other than the rollout, it will automatically heal back to the spec defined in the parent MonoVertexRollout.

Added unit tests to verify correctness.

Updated Dockerfile to go version 1.22 as required by the new Numaflow version v1.30-rc1 which contains the MonoVertex CRD.

Updated and added roles for permissions pertaining to the new CRDs.

### Verification

Creating MonoVertexRollout locally to test creation, deletion and updating manually.

Ran `make test` after creating tests as well.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -202,6 +202,7 @@ ifeq ($(NUMAFLOW_CRDS), 0)
 	$(KUBECTL) apply -f https://raw.githubusercontent.com/numaproj/helm-charts/main/charts/numaflow/crds/isbsvcs.yaml
 	$(KUBECTL) apply -f https://raw.githubusercontent.com/numaproj/helm-charts/main/charts/numaflow/crds/pipelines.yaml
 	$(KUBECTL) apply -f https://raw.githubusercontent.com/numaproj/helm-charts/main/charts/numaflow/crds/vertices.yaml
+	$(KUBECTL) apply -f https://raw.githubusercontent.com/numaproj/numaflow/main/config/base/crds/full/numaflow.numaproj.io_monovertices.yaml
```
Comment (afugazzotto) [inline Makefile:205]: I don't think `numaflow-crd` is used at all, so we may not need to add this here.

File: config/crd/kustomization.yaml [hunk 2]
```diff
@@ -5,5 +5,6 @@ resources:
 - bases/numaplane.numaproj.io_pipelinerollouts.yaml
 - bases/numaplane.numaproj.io_numaflowcontrollerrollouts.yaml
 - bases/numaplane.numaproj.io_isbservicerollouts.yaml
+- bases/numaplane.numaproj.io_monovertexrollouts.yaml
 # install numaflow minimal CRDs as a dependency
 - https://github.com/numaproj/numaflow/config/advanced-install/minimal-crds?ref=stable
```
Comment (afugazzotto) [inline config/crd/kustomization.yaml:10]: Add Numaflow MonoVertex CRD here

Comment (dpadhiar) [reply config/crd/kustomization.yaml:10]: The minimal CRD for MonoVertex is already included here: https://github.com/numaproj/numaflow/tree/main/config/base/crds/minimal

Perhaps all we need to do is remove Numaflow CRDs from the Makefile then.

File: internal/controller/monovertexrollout_controller.go [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+
+func (r *MonoVertexRolloutReconciler) ErrorHandler(monoVertexRollout *apiv1.MonoVertexRollout, err error, reason, msg string) {
+	// r.customMetrics.MonoVerticesSyncFailed.WithLabelValues().Inc()
+	r.recorder.Eventf(monoVertexRollout, corev1.EventTypeWarning, reason, msg+" %v", err.Error())
+}
+
+func getMonoVertexChildResourceHealth(conditions []metav1.Condition) (metav1.ConditionStatus, string) {
+	for _, cond := range conditions {
+		switch cond.Type {
+		case "DaemonHealthy", "PodsHealthy":
```
Comment (xdevxy) [inline internal/controller/monovertexrollout_controller.go:354]: You confirmed there are only two conditions for monovertex? Thanks!

File: internal/controller/monovertexrollout_controller.go [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+			if err := kubernetes.CreateCR(ctx, r.restConfig, newMonoVertexDef, "monovertices"); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			monoVertexRollout.Status.MarkDeployed(monoVertexRollout.Generation)
+		} else {
+			return ctrl.Result{}, fmt.Errorf("error getting MonoVertex: %v", err)
+		}
+	} else {
+		// merge and update
```
Comment (xdevxy) [inline internal/controller/monovertexrollout_controller.go:196]: Can you add a TODO saying for monovertex we can directly apply changes and no need to do draining?

File: config/samples/numaplane.numaproj.io_v1alpha1_numaflowcontrollerrollout.yaml [hunk 5]
```diff
@@ -5,4 +5,4 @@ metadata:
   namespace: example-namespace
 spec:
   controller:
-    version: "0.0.7" #numaio-rc-0.0.7
+    version: "1.3.0-rc1" #numaio-rc-0.0.7, 1.3.0-rc1
```
[the remark this answers is not in the document]
Comment (dpadhiar) [reply config/samples/numaplane.numaproj.io_v1alpha1_numaflowcontrollerrollout.yaml:8]: Not needed now if we are okay with leaving `1.3.0-rc1` as default, will remove.

File: internal/controller/monovertexrollout_controller.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.3/pkg/reconcile
+func (r *MonoVertexRolloutReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	// syncStartTime := time.Now()
+	numaLogger := logger.GetBaseLogger().WithName("monovertexrollout-reconciler").WithValues("monovertexrollout", req.NamespacedName)
+
+	// update the context with this Logger
+	ctx = logger.WithLogger(ctx, numaLogger)
+	// r.customMetrics.MonoVerticesSynced.WithLabelValues().Inc()
```
[the remark this answers is not in the document]
Comment (dpadhiar) [reply internal/controller/monovertexrollout_controller.go:93]: Hadn't added all the implementation for metrics yet when I was writing this.

Could remove for now for a separate metrics PR or include in this one.

File: internal/controller/monovertexrollout_controller.go [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+		newMonoVertexDef = mergeMonoVertex(existingMonoVertexDef, newMonoVertexDef)
+		err := r.updateMonoVertex(ctx, monoVertexRollout, newMonoVertexDef)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	if err = r.applyPodDisruptionBudget(ctx, monoVertexRollout); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to apply PodDisruptionBudget for MonoVertexRollout %s, err: %v", monoVertexRollout.Name, err)
+	}
```
[the remark this answers is not in the document]
Comment (dpadhiar) [reply internal/controller/monovertexrollout_controller.go:206]: If this is the case, will remove!

File: internal/controller/monovertexrollout_controller.go (hunk 3 above)
Comment (dpadhiar) [reply internal/controller/monovertexrollout_controller.go:354]: Yes, these are the only conditions for subresource health. The other condition a MonoVertex will have is `Deployed`.

File: internal/controller/monovertexrollout_controller.go [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+		numaLogger.Errorf(err, "failed to parse status from MonoVertex: %+v, %v", monoVertex, err)
+		return
+	}
+
+	numaLogger.Debugf("monoVertex status: %+v", monoVertexStatus)
+
+	monoVertexPhase := numaflowv1.MonoVertexPhase(monoVertexStatus.Phase)
+	monoVertexChildResourceStatus, monoVertexChildResourceReason := getMonoVertexChildResourceHealth(monoVertexStatus.Conditions)
+
+	if monoVertexChildResourceReason == "Progressing" {
```
[the remark this answers is not in the document]
Comment (dpadhiar) [reply internal/controller/monovertexrollout_controller.go:286]: Not that I can see, if it doesn't should we not check for `Progressing`?

These are the only Phases defined for MonoVertex:
```
MonoVertexPhaseUnknown MonoVertexPhase = ""
MonoVertexPhaseRunning MonoVertexPhase = "Running"
MonoVertexPhaseFailed  MonoVertexPhase = "Failed"
```

File: internal/controller/monovertexrollout_controller.go (hunk 4 above)
Comment (dpadhiar) [reply internal/controller/monovertexrollout_controller.go:196]: Can do, for `apply change directly` what should the logic be?

File: internal/controller/monovertexrollout_controller.go [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+//+kubebuilder:rbac:groups=numaplane.numaproj.io,resources=monovertexrollouts/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.3/pkg/reconcile
+func (r *MonoVertexRolloutReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	// syncStartTime := time.Now()
```
[the remark this answers is not in the document]
Comment (dpadhiar) [reply internal/controller/monovertexrollout_controller.go:88]: Same reasoning as metrics line comment.

File: internal/controller/monovertexrollout_controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (xdevxy) [reply internal/controller/monovertexrollout_controller.go:285]: no just ensuring we do not lose the pausing state if present. as they do not have it for now, this is good.

File: internal/controller/monovertexrollout_controller.go (hunk 4 above)
Comment (xdevxy) [reply internal/controller/monovertexrollout_controller.go:194]: this is what you have right, just apply the spec directly without checking if needs pausing/draining.

Comment (xdevxy) [reply internal/controller/monovertexrollout_controller.go:194]: oh my bad, not TODO, but a comment... sorry for the typo

**Review by xdevxy — APPROVED** [review]

**Pull request merged**